src/Microsoft.ML.Recommend/MatrixFactorizationPredictor.cs

```diff
@@ -0,0 +1,403 @@
+//------------------------------------------------------------------------------
+// <copyright company="Microsoft Corporation">
+//     Copyright (c) Microsoft Corporation.  All rights reserved.
+// </copyright>
+//------------------------------------------------------------------------------
+
+using Float = System.Single;
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using Microsoft.ML.Runtime;
+using Microsoft.ML.Runtime.Data;
+using Microsoft.ML.Runtime.Internal.Internallearn;
+using Microsoft.ML.Runtime.Internal.Utilities;
+using Microsoft.ML.Runtime.Model;
+using Microsoft.ML.Runtime.Recommend;
+using Microsoft.ML.Runtime.Recommend.Internal;
+
+[assembly: LoadableClass(typeof(MatrixFactorizationPredictor), null, typeof(SignatureLoadModel),
+    "Matrix Factorization Predictor Executor", MatrixFactorizationPredictor.LoaderSignature)]
+
+namespace Microsoft.ML.Runtime.Recommend
+{
+    public sealed class MatrixFactorizationPredictor : IPredictor, ICanSaveModel, ICanSaveInTextFormat, IMIValueMapper, ISchemaBindableMapper,
+        IUserHistoryToItemsRecommender
+    {
+        public const string LoaderSignature = "MFPredictorExec";
+        public const string RegistrationName = "MatrixFactorizationPredictor";
+
+        private static VersionInfo GetVersionInfo()
+        {
+            return new VersionInfo(
+                "MFPREDIC",
+                0x00010001, // Initial
+                0x00010001,
+                0x00010001,
+                "MFPredictorExec",
+                null);
+        }
+
+        private readonly IHost _host;
+        // The number of rows.
+        private readonly int _m;
+        // The number of columns.
+        private readonly int _n;
+        // The internal dimension.
+        private readonly int _k;
+        // The libMF is always single precision, so we will
+        // keep it as such even in the double build, even when
+        // the TLC output type is noted as being double.
+
+        // Packed _m by _k matrix.
+        private readonly Single[] _p;
```

Comment:
>Single [](start = 25, length = 6)

If Single is used directly - is the Float declaration above needed?

Reply:
Float and Single will be replaced by float. Thanks.

---
In reply to: [225353234](https://github.com/dotnet/machinelearning/pull/1263#discussion_r225353234) [](ancestors = 225353234)